### PR TITLE
feat: Support text codec

### PR DESCRIPTION
**Pull request by neko-para**
## Description
Support basic guessing codec.
Support changing saving codec.

## Related Issue
#319

## Type of changes
- [x] New feature (changes which add functionality)

### CODE REVIEW

**Comment by neko-para** [comment]
Actually state.invalidChars doesn't work perfectly.
Currently, we can put system and specified codec(like gb2312) before utf-8, as these codecs are generally more strict.
Therefore, a better solution is needed.

**Comment by ouuan** [comment]
> Actually state.invalidChars doesn't work perfectly.
> Currently, we can put system and specified codec(like gb2312) before utf-8, as these codecs are generally more strict.
> Therefore, a better solution is needed.

In this case, I'll convert this PR to a draft PR, and you can open it for review when it's ready.

**Comment by neko-para** [comment]
I've got up an idea. We can add a status bar which contains the info and buttons like vscode, and the codepage and line ending could also be added there.

File: src/Util.cpp [hunk 1]
```diff
@@ -105,20 +112,17 @@ QString readFile(const QString &path, const QString &head, MessageLogger *log, b
                 log->warn(head, QString("The file [%1] does not exist").arg(path));
             LOG_WARN(QString("The file [%1] does not exist").arg(path));
         }
-        return QString();
+        return "";
```
Comment (ouuan) [inline src/Util.cpp:115]: I think you didn't read this?

https://github.com/cpeditor/cpeditor/blob/6565bb5c6fa5369b2b8528785ff35bc96b4073b0/src/Util.hpp#L54-L56

There seems to be no specific reason for changing this line.

File: src/Util.hpp [hunk 2]
```diff
@@ -18,6 +18,7 @@
 #ifndef UTIL_HPP
 #define UTIL_HPP
 
+#include <QByteArray>
```
Comment (ouuan) [inline src/Util.hpp:21]:
```suggestion
```

It compiles without this include on my machine.

**Comment by ouuan** [comment]
In fact, I don't think this feature is very useful :thinking: In #319, I'd rather suggest @2740365712 not to write comments in Chinese, or not to open a file with Chinese comments with both CP Editor and Dev C++. However, I'll accept your PR if there are no problems.

File: src/Util.cpp (hunk 1 above)
Comment (neko-para) [reply src/Util.cpp:115]: I noticed. But I think that QByteArray is the file content, and we convert it to QString by QString::fromUtf8, so I add the fromUtf8 or codec->toUnicode() to all the places used this function

**Comment by neko-para** [comment]
I know why you crashes. I'll fix it in next commit.

File: src/Util.cpp (hunk 1 above)
Comment (neko-para) [reply src/Util.cpp:115]: And as the return type is QByteArray, it's better to return a "" than a QString()

Comment (ouuan) [reply src/Util.cpp:115]: Is a non-exist file distinguished from an empty file now?

Comment (ouuan) [reply src/Util.cpp:115]: Maybe the correct choice is `QByteArray()`?

Comment (ouuan) [reply src/Util.cpp:115]: And you should also modify the comments.

Comment (neko-para) [reply src/Util.cpp:115]: ok, I'll do it

**Pull request closed without merging**